File: Backend/taskservices/controller/logController.js
import express from "express";
import * as logService from "../service/logService.js";

const router = express.Router();

/**
 * POST /logs
 * Create a new activity log entry.
 * Body: { userId, userName, action, entity, entityId, details, ipAddress }
 */
router.post("/", async (req, res) => {
  res.json(await logService.addLog(req.body));
});

/**
 * GET /logs?page=1&size=20
 * Get all logs (requires JWT token in header).
 */
router.get("/", async (req, res) => {
  const { page = 1, size = 20 } = req.query;
  res.json(await logService.getLogs(page, size, req.headers["token"]));
});

/**
 * GET /logs/recent?limit=10
 * Get recent activity (no auth — used by dashboard).
 */
router.get("/recent", async (req, res) => {
  const { limit = 10 } = req.query;
  res.json(await logService.getRecentLogs(limit));
});

/**
 * GET /logs/stats
 * Get count of logs grouped by action type (for charts).
 */
router.get("/stats", async (req, res) => {
  res.json(await logService.getLogStats());
});

/**
 * GET /logs/user/:userId?page=1&size=20
 * Get logs for a specific user.
 */
router.get("/user/:userId", async (req, res) => {
  const { userId } = req.params;
  const { page = 1, size = 20 } = req.query;
  res.json(await logService.getLogsByUser(userId, page, size));
});

export default router;

File: Backend/gateway/controllers/logController.py
from fastapi import APIRouter
from typing import Optional
import httpx

router = APIRouter(prefix="/api/logs")

NODE_URL = "http://localhost:8002"

@router.post("")
async def create_log(body: dict):
    """Create a new activity log entry (called internally by other services)."""
    async with httpx.AsyncClient() as client:
        response = await client.post(NODE_URL + "/logs", json=body)
    return response.json()

@router.get("")
async def get_all_logs(page: int = 1, size: int = 20, token: Optional[str] = None):
    """Get all activity logs (admin only – pass JWT token header)."""
    async with httpx.AsyncClient() as client:
        response = await client.get(
            NODE_URL + f"/logs?page={page}&size={size}",
            headers={"token": token or ""}
        )
    return response.json()

@router.get("/recent")
async def get_recent_logs(limit: int = 10):
    """Get the most recent N activity logs (for dashboard widget)."""
    async with httpx.AsyncClient() as client:
        response = await client.get(NODE_URL + f"/logs/recent?limit={limit}")
    return response.json()

@router.get("/stats")
async def get_log_stats():
    """Get log counts grouped by action type (for analytics charts)."""
    async with httpx.AsyncClient() as client:
        response = await client.get(NODE_URL + "/logs/stats")
    return response.json()

@router.get("/user/{user_id}")
async def get_logs_by_user(user_id: int, page: int = 1, size: int = 20):
    """Get activity logs for a specific user."""
    async with httpx.AsyncClient() as client:
        response = await client.get(
            NODE_URL + f"/logs/user/{user_id}?page={page}&size={size}"
        )
    return response.json()
